### PR TITLE
Renovation: add onVisibilityChange callback functionality

### PR DESCRIPTION
**Pull request by MaximKudriavtsev**

### CODE REVIEW

**Review by LazyLahtak — DISMISSED** [review]

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{
+  // Use IntelliSense to learn about possible Node.js debug attributes.
+  // Hover to view descriptions of existing attributes.
+  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
```
Comment (LazyLahtak) [inline .vscode/launch.json:4]: Should we keep this?

Comment (MaximKudriavtsev) [reply .vscode/launch.json:4]: It auto-generated comment. I think it is not important

Comment (MaximKudriavtsev) [reply .vscode/launch.json:4]: Better to keep it

**Review by LazyLahtak — DISMISSED** [review]

File: js/renovation/widget.tsx [hunk 2]
```diff
@@ -143,7 +145,7 @@ export const viewFunction = (viewModel: any) => {
 export class WidgetInput {
     @OneWay() _feedbackHideTimeout?: number = 400;
     @OneWay() _feedbackShowTimeout?: number = 30;
-    @OneWay() _visibilityChanged?: (args: any) => undefined;
+    @OneWay() onVisibilityChange?: (args: any) => undefined;
```
Comment (churkin) [inline js/renovation/widget.tsx]: boolean argument?

File: js/renovation/widget.tsx [hunk 3]
```diff
@@ -190,13 +192,11 @@ export default class Widget extends JSXComponent<WidgetInput> {
 
     @Effect()
     visibilityEffect() {
-        const { name, _visibilityChanged, visible } = this.props;
-        const namespace = `${name}VisibilityChange`;
-
-        if (_visibilityChanged !== undefined && hasWindow()) {
+        const namespace = `${this.props.name}VisibilityChange`;
```
Comment (churkin) [inline js/renovation/widget.tsx]: const { name, onVisibilityChange } = this.props;

File: testing/jest/widget.tests.tsx [hunk 4]
```diff
@@ -362,6 +362,27 @@ describe('Widget', () => {
         });
     });
 
+    describe('Callbacks', () => {
+        describe('onVisibilityChanged', () => {
+            it('is called on `dxhiding` and `dxshown` events and special css class is attached', () => {
+                const onVisibilityChange = jest.fn();
+                const widget = render({ onVisibilityChange });
+
+                expect(widget.find('.dx-visibility-change-handler').exists()).toBe(true);
```
Comment (churkin) [inline testing/jest/widget.tests.tsx]: widget.exists('.dx-visibility-change-handler')

File: testing/jest/widget.tests.tsx [hunk 5]
```diff
@@ -362,6 +362,27 @@ describe('Widget', () => {
...
+
+                emit(EVENT.hiding);
+                widget.setProps({});
+                expect(onVisibilityChange).toHaveBeenCalledTimes(1);
+                expect(onVisibilityChange).toHaveBeenLastCalledWith(false);
+
+                emit(EVENT.shown);
+                widget.setProps({});
```
Comment (churkin) [inline testing/jest/widget.tests.tsx]: Can we remove this now?

**Review by churkin — CHANGES_REQUESTED** [review]

**Review by churkin — APPROVED** [review]

**Review by LazyLahtak — APPROVED** [review]

**Pull request merged**